### ISSUE TITLE
Implicitly bound route parameters are not always available inside middleware

### ISSUE THREAD

**Issue by jacovdbergh** [laravel/framework#45223]
- Laravel Version: v9.41.0
- PHP Version: 8.1
- Database Driver & Version: n/a

### Description:

When using implicit route model binding, Laravel looks at the controller methods arguments to infer the concrete types of the route parameters. When they exist, Laravel will automatically _convert_ the route parameter to an instance of the model.

Route parameters are not only used inside the routes controller, but might also be used inside middleware. A common example is authorisation.

If the controller does not have have all of the route's parameters declared as method parameters, Laravel won't convert the unmatched route parameters to models.

This means the Middleware (when using `request()->route()->parameter('x')`) receives a string, and not the intended model.

Let's look at an example:
```php
Route::domain('{tenant:domain}')
     ->get('/projects/{project:id}', ShowProjectController::class)
     ->middleware(VerifyUserHasAccessToTenantMiddleware::class);
```

```php
class ShowProjectController
{
    public function __invoke(Project $project)
    {
        // do something with $project
    }
}
```

```php
class VerifyUserHasAccessToTenantMiddleware
{
    public function handle(Request $request, Closure $next)
    {
        $tenant = $request->route()->parameter('tenant');
        // $tenant is either string or Tenant::class depending on the route's controller!
    }
}
```

When using implicit route model binding, calling `request()->route()->parameter('tenant')` from our `VerifyUserHasAccessToTenantMiddleware` middleware returns a string, because the model is not bound due to `$tenant` not being declared in the ShowProjectController class.

The easy fix would be to _just_ declare the variable on all the controllers, so the implicit binding will automatically bind it. However, you are now adding a variable to a controller, to make it available in a middleware. This feels like a smell. Additionally, this is not possible when using Livewire full page components.

Most developers would rather explicitly bind the variable (using `Route::model()`), so they won't have to contaminate their controllers with (potentially) unused variables.

Unfortunately, with that you now lose the ability to specify custom key names and to use scoped binding.

The documentation state the following:
> You are not required to use Laravel's implicit, convention based model resolution in order to use model binding. You can also explicitly define how route parameters correspond to models. To register an explicit binding, use the router's model method to specify the class for a given parameter.

Using `Route::model()` pretends to be only a change to the way the model class name is resolved, but it actually changes more behaviour.

Specifically, the following two things don't work anymore:
- Custom key names
- Scoped bindings

Scoped binding behaviour can easily be re-enabled using `->scopeBindings()`.
But custom key names can not be used at all.

## Potential solution

`Route::model()` should ideally only change the model name resolution logic, and still support custom key names and scoped bindings. Only when you go to _full_ explicit binding (using `Route::bind(...)`), you should lose these features as you now opted in to take full control.

Relates to #33079
Relates to #45214

**Comment by driesvints** [comment]
Hi @jacovdbergh. I appreciate the thorough issue but it looks like Taylor does not want to change anything here right now. Maybe it's best to attempt a PR to the docs? Thank you

**Comment by georgeboot** [comment]
@driesvints Judging from Taylor's reply on the PR (#45214), we got the impression he didn't get the full picture and it might have looked like a user error on our side.

That's why @jacovdbergh and I spent a full day crafting this issue so we could explain the exact issue in detail.

This is an actual issue that we have not hit again and decided we should rather fix it than to work around it. We have hit it many times before, and so have other developers. Also, Taylor actually asked for a PR in the older issue (#33079).

As the proposed solution doesn't have any real drawbacks (that we can see), mind giving a bit more insights into why a fix for this issue is not wanted at the moment?

This reply is not to get our way or similar, it's to get an understanding why the Laravel team does not consider this to be an issue, and how we can go forward from here.

Thank you for your valuable time!
